File: cocos/physics/physx/joints/physx-fixed-joint.ts
/*
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { IVec3Like, Vec3, Quat, Mat4 } from '../../../core';
import { FixedConstraint, PhysicsSystem } from '../../framework';
import { IFixedConstraint } from '../../spec/i-physics-constraint';
import { PX, _trans, getTempTransform, _pxtrans } from '../physx-adapter';
import { PxContactPairFlag } from '../physx-enum';
import { PhysXInstance } from '../physx-instance';
import { PhysXRigidBody } from '../physx-rigid-body';
import { PhysXWorld } from '../physx-world';
import { PhysXJoint } from './physx-joint';

const v3_0 = new Vec3();
const quat_0 = new Quat();
const mat_0 = new Mat4();

export class PhysXFixedJoint extends PhysXJoint implements IFixedConstraint {
    setBreakForce (v: number): void {
        this._breakForce = this.constraint.breakForce;
        this._impl.setBreakForce(this._breakForce, this._breakTorque);
    }

    setBreakTorque (v: number): void {
        this._breakTorque = this.constraint.breakTorque;
        this._impl.setBreakForce(this._breakForce, this._breakTorque);
    }

    get constraint (): FixedConstraint {
        return this._com as FixedConstraint;
    }

    private _breakForce = 0;
    private _breakTorque = 0;

    onComponentSet (): void {
        this._impl = PX.createFixedConstraint(PhysXJoint.tempActor, _pxtrans, null, _pxtrans);
        this.setBreakForce(this.constraint.breakForce);
        this.setBreakTorque(this.constraint.breakTorque);
        this.updateFrame();
        this.enableDebugVisualization(true);
    }

    updateFrame (): void {
        const bodyA = (this._rigidBody.body as PhysXRigidBody).sharedBody;
        const cb = this.constraint.connectedBody;

        Mat4.fromRT(mat_0, bodyA.node.worldRotation, bodyA.node.worldPosition);
        Mat4.invert(mat_0, mat_0);
        Mat4.getRotation(quat_0, mat_0);
        Mat4.getTranslation(v3_0, mat_0);
        this._impl.setLocalPose(0, getTempTransform(v3_0, quat_0));

        if (cb) {
            const bodyB = (cb.body as PhysXRigidBody).sharedBody;
            Mat4.fromRT(mat_0, bodyB.node.worldRotation, bodyB.node.worldPosition);
            Mat4.invert(mat_0, mat_0);
            Mat4.getRotation(quat_0, mat_0);
            Mat4.getTranslation(v3_0, mat_0);
            this._impl.setLocalPose(1, getTempTransform(v3_0, quat_0));
        } else {
            this._impl.setLocalPose(1, getTempTransform(Vec3.ZERO, Quat.IDENTITY));
        }
    }

    updateScale0 (): void {
        this.updateFrame();
    }

    updateScale1 (): void {
        this.updateFrame();
    }
}

File: native/cocos/physics/physx/joints/PhysXFixedJoint.cpp
/****************************************************************************
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "physics/physx/joints/PhysXFixedJoint.h"
#include "math/Quaternion.h"
#include "physics/physx/PhysXSharedBody.h"
#include "physics/physx/PhysXUtils.h"

namespace cc {
namespace physics {

void PhysXFixedJoint::onComponentSet() {
    _transA = physx::PxTransform(physx::PxIdentity);
    _transB = physx::PxTransform(physx::PxIdentity);

    physx::PxRigidActor *actor0 = _mSharedBody->getImpl().rigidActor;
    physx::PxRigidActor *actor1 = nullptr;

    if (_mConnectedBody) {
        actor1 = _mConnectedBody->getImpl().rigidActor;
    }

    _mJoint = PxFixedJointCreate(PxGetPhysics(), actor0, _transA, actor1, _transB);

    updatePose();
    setEnableDebugVisualization(true);
}

void PhysXFixedJoint::setBreakForce(float force) {
    _breakForce = force;
    _mJoint->setBreakForce(_breakForce, _breakTorque);
}

void PhysXFixedJoint::setBreakTorque(float torque) {
    _breakTorque = torque;
    _mJoint->setBreakForce(_breakForce, _breakTorque);
}

void PhysXFixedJoint::updateScale0() {
    updatePose();
}

void PhysXFixedJoint::updateScale1() {
    updatePose();
}

void PhysXFixedJoint::updatePose() {
    _transA = physx::PxTransform(physx::PxIdentity);
    _transB = physx::PxTransform(physx::PxIdentity);

    pxSetVec3Ext(_transA.p, _mSharedBody->getNode()->getWorldPosition());
    pxSetQuatExt(_transA.q, _mSharedBody->getNode()->getWorldRotation());
    if (_mConnectedBody) {
        pxSetVec3Ext(_transB.p, _mConnectedBody->getNode()->getWorldPosition());
        pxSetQuatExt(_transB.q, _mConnectedBody->getNode()->getWorldRotation());
    }
    _mJoint->setLocalPose(physx::PxJointActorIndex::eACTOR0, _transA.getInverse());
    _mJoint->setLocalPose(physx::PxJointActorIndex::eACTOR1, _transB.getInverse());
}

} // namespace physics
} // namespace cc
